gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/MfaFilterContext.java

```diff
@@ -131,6 +133,29 @@ public boolean isDeviceRiskAssessmentEnabled() {
     public boolean isChallengeOnceCompleted() {
         return TRUE.equals(session.get(MFA_CHALLENGE_COMPLETED_KEY));
     }
+
+
+    public void setDefaultFactorWhenApplied(RuleEngine ruleEngine) {
+        String enrollingFactorId = session.get(ENROLLED_FACTOR_ID_KEY);
+        FactorSettings factorSettings = client.getFactorSettings();
+        if (nonNull(enrollingFactorId) && nonNull(factorSettings) && !matchFactorRule(enrollingFactorId, factorSettings, ruleEngine)) {
+            session.put(ENROLLED_FACTOR_ID_KEY, getDefaultFactorId());
+        }
+    }
+
+    public boolean matchFactorRule(String factorId, FactorSettings factorSettings, RuleEngine ruleEngine) {
+        var appFactor = ofNullable(factorSettings.getApplicationFactors()).flatMap(factors -> factors.stream().filter(f -> f.getId().equals(factorId)).findFirst());
+        return appFactor.isPresent() && (factorSettings.getDefaultFactorId().equals(factorId) || isMatchFactorRule(ruleEngine, appFactor.get().getSelectionRule()));
+    }
+
+    public boolean isMatchFactorRule(RuleEngine ruleEngine, String selectionRule) {
```

Comment:
**suggestion**: RuleEngine as class attribute

Reply:
except if I'm wrong, this ruleEngine parameter is still present and doesn't rely on the class attribute.